portability-worker/src/main/java/org/dataportabilityproject/worker/JobPollingService.java

```diff
@@ -71,17 +71,20 @@ private void pollForUnassignedJob() throws IOException {
     String id = jobDao.findNextJobPendingWorkerAssignment();
     logger.debug("Polled pollForUnassignedJob, found id: {}", id);
     if (id != null) {
-      PortabilityJob job = jobDao.lookupJobPendingWorkerAssignment(id);
-      Preconditions.checkNotNull(job);
       Preconditions.checkState(!jobMetadata.isInitialized());
       jobMetadata.init(id);
-
       PublicKey publicKey = jobMetadata.getKeyPair().getPublic();
       // TODO: Move storage of private key to a different location
       PrivateKey privateKey = jobMetadata.getKeyPair().getPrivate();
       // Executing Job State Transition from Unassigned to Assigned
-      jobDao.updateJobStateToAssignedWithoutAuthData(id, publicKey, privateKey);
-      logger.debug("Completed updateJobStateToAssignedWithoutAuthData, publicKey: {}", publicKey.getEncoded().length);
+      if (jobDao.updateJobStateToAssignedWithoutAuthData(id, publicKey, privateKey)) {
+        logger.debug("Completed updateJobStateToAssignedWithoutAuthData, publicKey: {}",
+            publicKey.getEncoded().length);
+      } else {
+        logger.debug("Failed to claim job {}; it was probably already claimed by another worker",
+            id);
+        jobMetadata.reset();
```

Comment:
Instead of reset. Let's create the public and private key outside of the jobMetatData.

Then we update the method call jobMetadata.init(id, publicKey, privateKey) if we're successful in assigning the.

Another option is to process the assignment update then, when that succeeds, update with the key info.

Thoughts?

Reply:
Good idea, done